core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -955,6 +940,80 @@ public void run() {
       return pendingCall;
     }
 
+    // Must run in SynchronizationContext.
+    private void drainPendingCalls() {
+      if (pendingCalls == null) {
+        return;
+      }
+      for (RealChannel.PendingCall<?, ?> pendingCall : pendingCalls) {
+        pendingCall.reprocess();
+      }
+    }
+
+    void shutdown() {
+      if (!shutdown.compareAndSet(false, true)) {
+        return;
+      }
+      // Put gotoState(SHUTDOWN) as early into the syncContext's queue as possible.
+      // delayedTransport.shutdown() may also add some tasks into the queue. But some things inside
+      // delayedTransport.shutdown() like setting delayedTransport.shutdown = true are not run in
+      // the syncContext's queue and should not be blocked, so we do not drain() immediately here.
+      syncContext.executeLater(new Runnable() {
+        @Override
+        public void run() {
+          channelLogger.log(ChannelLogLevel.INFO, "Entering SHUTDOWN state");
+          channelStateManager.gotoState(SHUTDOWN);
+        }
+      });
+      if (configSelector.get() != INITIAL_PENDING_SELECTOR && pendingCalls == null) {
```

Comment:
Why have this block of code? Won't the Shutdown Runnable do the same thing? Ditto for shutdownNow.

Reply:
This is to keep the fast path as https://github.com/grpc/grpc-java/blob/v1.32.0/core/src/main/java/io/grpc/internal/ManagedChannelImpl.java#L791

I don't mind if delete the fast path. Could there be any impact?

Reply:
I don't understand. Why do we need a fast path? Why do you think the old path was a "fast path"? Won't the "slow" path be virtually the same speed, although it could be delayed a bit in time.

Reply:
The only difference between "fast path" and  "slow path" is the former may eagerly set `delayedTransport.shutdown = true`. I think it's pretty safe to remove the fast path.

Removed.